cmd/mtaBuild.go

```diff
@@ -0,0 +1,411 @@
+package cmd
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/SAP/jenkins-library/pkg/command"
+	piperhttp "github.com/SAP/jenkins-library/pkg/http"
+	"github.com/SAP/jenkins-library/pkg/log"
+	"github.com/SAP/jenkins-library/pkg/piperutils"
+	"github.com/SAP/jenkins-library/pkg/telemetry"
+	"gopkg.in/yaml.v2"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"text/template"
+	"time"
+)
+
+const templateMtaYml = `_schema-version: "2.0.0"
+ID: "{{.ID}}"
+version: {{.Version}}
+
+parameters:
+  hcp-deployer-version: "1.0.0"
+
+modules:
+  - name: {{.Name}}
+    type: html5
+    path: .
+    parameters:
+       version: {{.Version}}-${timestamp}
+    build-parameters:
+      builder: grunt
+      build-result: dist`
+
+// MTABuildTarget ...
+type MTABuildTarget int
+
+const (
+	// NEO ...
+	NEO MTABuildTarget = iota
+	// CF ...
+	CF MTABuildTarget = iota
+	//XSA ...
+	XSA MTABuildTarget = iota
+)
+
+// ValueOfBuildTarget ...
+func ValueOfBuildTarget(str string) (MTABuildTarget, error) {
+	switch str {
+	case "NEO":
+		return NEO, nil
+	case "CF":
+		return CF, nil
+	case "XSA":
+		return XSA, nil
+	default:
+		return -1, fmt.Errorf("Unknown BuildTarget: '%s'", str)
+	}
+}
+
+// String ...
+func (m MTABuildTarget) String() string {
+	return [...]string{
+		"NEO",
+		"CF",
+		"XSA",
+	}[m]
+}
+
+func mtaBuild(config mtaBuildOptions, telemetryData *telemetry.CustomData, commonPipelineEnvironment *mtaBuildCommonPipelineEnvironment) {
+	log.Entry().Info("Launching mta build")
+	err := runMtaBuild(config, commonPipelineEnvironment, &command.Command{})
+	if err != nil {
+		log.Entry().
+			WithError(err).
+			Fatal("failed to execute mta build")
+	}
+
+}
+
+func runMtaBuild(config mtaBuildOptions, commonPipelineEnvironment *mtaBuildCommonPipelineEnvironment,
+	e envExecRunner) error {
+
+	e.Stdout(os.Stderr) // keep stdout clear.
```

Comment:
please direct logging to logrus, see for example here: https://github.com/SAP/jenkins-library/blob/master/cmd/karmaExecuteTests.go#L15

Reply:
@OliverNocon: Can you explain more about log forwarding  via `logrus`.

I have still the concern that we end up in log messages like `[INFO][ERROR] build failed`. These kind of log message are very confusing and hard to handle by some automated log scanners.

Reply:
When implementing the first step with command and logging we discussed this topic in detail and came to the conclusion to do it as implemented in e.g. `karmaExecuteTests`.

Yes, there is the downside that depending on the tooling we see messages like `[INFO][ERROR]  <something>` in the console.

On the other side we have several downsides when not going through `logrus` in all cases:

#### 1. additional log info missing

Logrus will provide us with additional information like timestamp, stepName, library, ...
This is helpful in case of issues since important info is available right away and we do not depend on the CI/CD system and its configuration (e.g. timestamp plugin in Jenkins).

#### 2. Not being able to use the `logrus` functionality to forward logs to e.g. LogStash.

Forwarding will typically be done in a JSON format (main info contained in the `message` field) and as such the log message will appear in the respective logging system as provided by `mvn`, `npm`, ...

#### 3. Inconsistencies across steps

Steps so far use `logrus`, unless we overlooked s.th. then we have to adapt it there as well.

If we don't use `logrus` for some steps output (in e.g. LogStash) will just be silent if there is no access to the console ...

**We considered these aspects and took the concious decision to go the `logrus` route accepting that in certain cases messages in the console are not 100% ideal.**
This is reflected in the way of implementing the step `karmaExecuteTests` as well as other steps by now.

If we see severe issues with the current approach among end users we still are able to influence the formatting of the log in the console to improve readability.
With `logrus` we can do this centrally without any impact on the step implementation.